crystal/src/y3storage/autoinst_issues/exception.cr

```diff
@@ -0,0 +1,69 @@
+# Copyright (c) [2019-2020] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "./issue"
+
+module Y3Storage
+  module AutoinstIssues
+    # Represents a problem that occurs when an exception is raised.
+    #
+    # This error is used as a fallback for any problem that arises during
+    # proposal which is not handled in an specific way. It includes the
+    # exception which caused the problem to be registered.
+    #
+    # Example: registering an exception
+    #     begin
+    #       do_stuff # some exception is raised
+    #     rescue e : SomeException
+    #       new Y3Storage::AutoinstIssues::Exception.new(e)
+    #     end
+    class Exception < Issue
+      # Exception that was rescued
+      getter error : ::Exception
+
+      # Constructor
+      #
+      # Gets only one argument of type `::Exception`
+      def initialize(*args)
+        first = args[0]?
+        if first && first.is_a?(::Exception)
+          @error = first
+        else
+          raise ArgumentError.new("Wrong initialization: #{args}")
+        end
```

Comment:
maybe FIXME here as I expect crystal can catch it during runtime?

Reply:
I don't think it's doable due to the design of the original Ruby code.

In that Ruby code, each subclass of `Issue` has a completely different constructor with a completely different set of arguments. And then, new objects are instantiated as:

```ruby
def create_issue(type, *extra_args)
  class_name = type.to_s.split("_").map(&:capitalize).join
  klass = Y2Storage::AutoinstIssues.const_get(class_name)
  klass.new(*extra_args)
end
```

We translated that in Crystal like this:

```crystal
def create_issue(type, *extra_args)
  klass = Issue.subclass(type)
  if klass
    klass.new(*extra_args)
  else
    raise Error.new("Unknown AutoInstIssue type")
  end
end
```

And that only compiles if the constructors of all the classes returned by `Issue.subclass` always have the same signature, which can only be `initialize(*args)`.

That forces us to implement the validation of argument on each subclass.